### ISSUE TITLE
Memory used keeps arising during training.

### ISSUE THREAD

**Issue by tppppppppp** [apache/incubator-mxnet#8011]
When I'm trying to train my network with python, the gpu memory used keeps arising during training. The network  is quite simple. As I'm a novice of mxnet, did I make some obvious mistake in my code?
Network definition:
![image](https://user-images.githubusercontent.com/13367895/30780961-134edb44-a149-11e7-99f9-dd96c0b4edad.png)

Training stage:
![image](https://user-images.githubusercontent.com/13367895/30780978-6469055e-a149-11e7-905b-13df83a68a5c.png)

Error Info:
![image](https://user-images.githubusercontent.com/13367895/30780986-9f8a0304-a149-11e7-8bc1-14648927bedc.png)

**Comment by piiswrong** [comment]
You need to synchronize at the end of each iteration.

for example, replace curr_loss += loss with curr_loss += loss.asnumpy()

**Comment by piiswrong** [comment]
Also looks like you are doing multiple backward between stepping. This wouldn't work unless you set grad_req to 'add' and use zero_grads after stepping

**Comment by tppppppppp** [comment]
Thanks. The gpu memory is steady after replacing curr_loss += loss with curr_loss += loss.asnumpy(). It looks as if the gpu memory of 'loss' is released after calling asnumpy. Would it not be realeased if I didn't call asnumpy? @piiswrong

**Comment by zhreshold** [comment]
@tppppppppp If you don't call asnumpy() explicitly here. The asynchronous feature make you worker trying to load the next batch before finishing the current batch. Thus may taken too much memory.

**Comment by tppppppppp** [comment]
Get it. Thanks a lot! @zhreshold

**Comment by eric-haibin-lin** [comment]
@tppppppppp has this issue been resolved? 

BTW there's an [example](https://github.com/apache/incubator-mxnet/blob/master/example/sparse/linear_classification.py) for sparse linear regression on cpu using the sparse ndarray feature in MXNet using the module API

**Comment by tppppppppp** [comment]
Yes. It has been resolved. Thanks for your example. @eric-haibin-lin
